feat(shared-data): json protocol schema 8

sfoster1 opened the pull request:
Add JSON protocol schema 8, which is similar in semantic concept to json protocol schema 7 but defines new keys for many large components (commands, annotations, liquids, labware) that set the expected schema for the content for that component.

For instance, now you specify a `commandSchemaId` and the elements of the `commands` array have to match that. You specify a `liquidSchemaId` and the elements of the `liquid` object have to match that.

This hopefully will let us be more flexible with upgrading things like the kinds of commands a protocol can run without having to change the protocol schema itself. Consumers will have to make sure that they can handle the schema for the components, but there's a new provided typescript validator (in `shared-data/js/protocols.ts`) that will validate any version of the protocol we support and return it with appropriate types; that's a bit more spread around in python but it similarly handles everything.

Also, you specify the deck schema version now.

This should be integrated into the robot for execution and we should make sure the app doesn't choke on it, but really it's a good way to be able to add new commands without making other protocols accidentally invalid.

This should be a big step forward in flexibility for our protocol schemas.

## testing
- [x] odd can render protocols tagged jp8
- [x] app can render protocols tagged jp8
- [x] offline analysis can handle protocols tagged jp8
- [x] robot analysis can handle protocols tagged jp8
- [x] robot can execute protocols tagged jp8
- [x] if you have a jp8 protocol run, analysis, and protocol saved on the robot and then you downgrade, the server is still operational

sfoster1 committed:
feat(shared-data): json protocol schema 8

Add JSON protocol schema 8, which is similar in semantic concept to json
protocol schema 7 but
- Requires user specification of the schema for the command List
This allows json protocols to switch which commands they're interning
without requiring further bumps to the protocol schema
- Does the same thing for command annotations, liquids, and labware
For the same reasons.

This should be a big step forward in flexibility for our protocol schemas.

b-cooper commented on shared-data/command-annotation/schemas/1.json:
This directory should be camelCase like the reference in the v8 schema `commandAnnotation/schemas`

b-cooper committed:
add enum to annotations and structure types for commands etc

b-cooper committed:
fix up ts config for protocol types

sfoster1 committed:
add python types and fixup some little schema things

sfoster1 committed:
add support in json_translator

SyntaxColoring commented on shared-data/protocol/schemas/8.json:
```diff
@@ -0,0 +1,161 @@
+
+  "type": "object",
+  "additionalProperties": false,
+  "required": [
+    "$otSharedSchema",
+    "schemaVersion",
+    "metadata",
+    "robot",
+    "labwareDefinitions",
+    "labwareDefinitionSchemaId",
+    "commands",
+    "commandsSchemaId",
+    "commandAnnotations",
+    "commandAnnotationsSchemaId",
+    "liquids",
+    "liquidSchemaId"
```
Hmm, I'm not sure about decoupling the `liquids` and `labwareDefinitions` schemas like this.

The argument for doing this for `commands` and `commandAnnotations` is that we expect them to change much more frequently than the other things in these files. Do we expect `liquids` and `labware` to change that frequently?

The downside that I have in mind is basically just complexity from the indirection.

sfoster1 replied:
labware, fair enough, but yeah I expect liquids to change pretty frequently as we start to actually like do stuff with them

jbleon95 commented:
As discussed in the comments on 13854, we should add some sort of `deckVersion` field along with `deckId` to allow robot server/protocol runner to know whether to load fixed trash or not, rather than just relying on schema version moving forward.

jerader commented on shared-data/protocol/schemas/8.json:
```diff
@@ -0,0 +1,161 @@
+      "items": { "$ref": "#/properties/commandSchemaRef" }
+    },
+
+    "commandAnnotationSchemaId": {
+      "description": "The schema to which the command annotations adhere.",
+      "type": "string"
+    },
+
+    "commandAnnotations": {
+      "type": "array",
+      "items": { "$ref": "#/properties/commandAnnotationSchemaRef" }
+    }
+  }
```
what is an example of `commandAnnotations`?

sfoster1 replied:
Something like grouping commands into the higher-level (i.e. PD, or python api call) command that created them, for instance. Or into user-specified arbitrary groups. Or time estimates, or really anything

sfoster1 committed:
update example files for better tests

sfoster1 committed:
update reader errors and add support to runner

sfoster1 committed:
add explicit ability to use dev server for tavern

If you are running a dev server, tavern will use that for its tests and
print out that it did so. You can also specify host and port for dev
server now.

sfoster1 committed:
this little guy needs a change

sfoster1 committed:
add new shared-data to tsconfig-eslint

sfoster1 committed:
export types as types from protocol

sfoster1 committed:
install new schemas on robot

sfoster1 committed:
format py

sfoster1 committed:
add a deck schema ID to protocols

sfoster1 committed:
don't try and use jsonschema for this

Jsonschema is a lot of things but one thing it isn't is a definer of
schema behavior based on values in the document. That's the key to all
of this - the document specifying a subschema for part of its content -
so we're just going to do it in code instead.

That means
- In the protocol schema, dynamically subschema'd objects get typed as
objects
- We get a new typescript validator to handle it all
- Also update python for this but it matters a bit less because the
python defines what is acceptable anyway, so there's no point trying to
be future proof here when downstream consumers will have to change

sfoster1 committed:
lint

sfoster1 committed:
js lint

sfoster1 committed:
py fixups

sfoster1 committed:
py format

jerader commented on shared-data/command/types/calibration.ts:
```diff
@@ -0,0 +1,84 @@
+  LabwareOffset,
+  Coordinates,
+} from '../../js/types'
+// TODO (sb 10/26/22): Separate out calibration commands from protocol schema in RAUT-272
```
I think we can delete this TODO, cc @smb2268

jerader commented on shared-data/js/protocols.ts:
```diff
@@ -0,0 +1,287 @@
+import type * as ProtocolSchemas from '../protocol'
+import type { CreateCommand } from '../command/types'
+import type { CommandAnnotation } from '../commandAnnotation/types'
+import Ajv from 'ajv'
```
can we move the Ajv import to be at the top of the imports? And then move the `type` imports to be at the bottom of the imports before line 22?

sfoster1 committed:
one last py format

jerader commented on shared-data/js/protocols.ts:
```diff
@@ -0,0 +1,287 @@
+  })
+
+// note: rejects with an array of ajv errors
+export function validate(
```
Maybe we can leave a note in here saying this is used in the `protocolValidation.test` since other files in the `shared-data/js` directory contain utils used outside of tests.

sfoster1 replied:
Oh this definitely will be used outside of tests, I just don't want to gate merging this pr on "switching the entire js ecosystem over to json protocol v8". The idea with this function is that you'll use it to validate protocols in js anywhere they appear since it will do the typecasting for you.

jerader reviewed (approved):
I reviewed the js stuff and sprinkled in a few clean up suggestions & notes. But lgtm! 🎉

sfoster1 committed:
review requests

SyntaxColoring commented on robot-server/tests/integration/conftest.py:
```diff
@@ -91,6 +91,14 @@ def _ot3_session_server(server_temp_directory: str) -> Generator[str, None, None
     base_url = (
         f"{_SESSION_SERVER_SCHEME}{_SESSION_SERVER_HOST}:{_OT3_SESSION_SERVER_PORT}"
     )
+
+    with _requests_session() as requests_session:
+        has_dev_server = _check_is_ready(requests_session, base_url)
+    if has_dev_server:
+        print("Using pre-existing flex dev server")
+        yield base_url
+        return
+
```
Is this intended to let you have `make -C robot-server dev` open in a second terminal, and if you run `make -C robot-server test`, it re-uses it? 

If so, that seems surprising. I expect tests to be isolated. If the tests can't run isolated because the port is already taken by a dev server, I expect them to error out. Could we leave this out?

Separately: if we did want to keep this in, we should do it consistently between the OT-2 and Flex dev servers.

sfoster1 replied:
yeah i can remove it

sfoster1 committed:
revert conftest changes

SyntaxColoring commented on shared-data/errors/definitions/1/errors.json:
```diff
@@ -205,6 +205,10 @@
     "4005": {
       "detail": "Command Parameter Limit Violated",
       "category": "generalError"
+    },
+    "4006": {
+      "detail": "Invalid Protocol Data",
+      "category": "generalError"
```
Nice, thank you.

SyntaxColoring reviewed (approved):
Python changes look great. Thank you!

b-cooper reviewed (approved):
🏈

sfoster1 committed:
bump the robot-server timeout
